### DISCUSSION TITLE
Google Calendar and Tasks sync

### DISCUSSION THREAD

**Discussion by mikeberger** [mikeberger/borg_calendar#155] [General]
BORG 1.10 adds the ability to sync with Google Calendar. 

Highlights:
- Non-todo events in BORG sync to google calendar events
- Todos in BORG sync to Google Tasks. Google Tasks only contain the BORG todo text and due date. They do not have the other todo attributes from BORG.
- aCalendar+ (android calendar) has been tested with BORG. aCalender+ adds repeated task functionality on top of google tasks. This works with BORG.
- Everything is free. No third party software is required to sync BORG and google calendar and android calendars.

Issues:
- Google Tasks are very primitive. BORG repeated tasks do not repeat in google. aCalendar+ adds repeating tasks.

And finally:
- Access to google calendar requires special google credentials for BORG. I cannot ship application credentials with BORG as they must remain secured (impossible in a Java app). 
- Other apps in this situation provide instructions to the users on how to register the app for themselves on the google cloud platform This seems unusual to me, but I have found a number of examples of this.

TLDR;
There is a google calendar sync in BORG 1.10. It is better than CALDAV which has serious issues due to unsupported 3pp libraries. However, to use it, the user must generate credentials via the google cloud platform.

**Comment by mikeberger** [comment]
### Google APIs and Security

Are the google task and calendar APIs open APIs? not really
Are they free? not really

So, you can't just enter your google credentials and start syncing BORG with google. It's just not that simple. 
Google requires the application (BORG) to have its own credentials. There are quotas for using these APIs - like number of queries per minute across ALL app users. 
The general idea is that if I were selling BORG and got lots of users, I'd have to pay google to increase the API quotas as my user's activity grows. That's how google would make money off of the APIs.

In BORG's google sync options, there is something called the Credentials File. This file isn't the user's credentials, it's BORG's credentials that I registered with google to uniquely identify BORG.

Well, I don't plan to pay google for higher quotas for free software, and I can't give out the credentials file anyway. It's supposed to be secret. 

**So, what to do about the app credentials?**

You would literally have to create and register an app in your name with google, tell google that your app uses the Tasks and Calendar APIs, get credentials for your app, and load them into Borg. 

Is this allowed?

I actually found more than 1 commercial app online telling their customers to do this. They just provide instructions on how to do it as if it's a normal thing for a user to do. 

Here is an example: https://simplyscheduleappointments.com/guides/google-api-credentials/

For BORG, this is what you'd technically have to do, except you'd need to add the Tasks API as well, and you'd have to download the credentials file when you are done.

**Comment by mikeberger** [comment]
So, if you want to play with the google sync, here's what you can do.

First, be aware that BORG is the master. The google sync is intended for BORG users that want to interact with their calendars on their phones or via ther web. It is not intended for google calendar users who want to view their google calendars in a desktop client. Just use a browser for that.

1. Backup anything in BORG and google that you don't want to lose. 
2. Create a new calendar and tasklist in google. For example, if you create a new calendar called "borgevents" and a new task list called "borgtasks", then these are the values to put in the BORG preferences under the Google tab. 
3. The BORG Google preferences also need a Credentials File and a Tokens Dir. For Tokens Dir, select any dir. I suggest making a new dir somewhere. It will just hold a cached google file.

The Credentials File

I am not going to maintain instructions on how to create this. Here are some examples on the internet:
https://simplyscheduleappointments.com/guides/google-api-credentials/
https://docs.simplecalendar.io/google-api-key/
https://docs.oracle.com/cloud-machine/latest/intcs_gs/ICSGO/GUID-941D575D-0B8B-4900-95ED-91D732E3B4AB.htm#ICSGO-GUID-941D575D-0B8B-4900-95ED-91D732E3B4AB

I suggest this one:

https://docs.simplecalendar.io/google-api-key/

If you follow the instructions from the link above, make the following changes:

- Repeat steps 6 and 7 for the Google Tasks API. BORG uses both the calendar and tasks API

- For step 12, create an OAuth client ID, not the API key

After step 12, there should be a way to download the OAuth credentials file. There should be a download icon to "download OAuth client". Save the downloaded file somewhere and put that file in the BORG google preferences.

At this point, if you have enabled google sync from the BORG preferences and filled in all the fields correctly, then when you press the BORG sync icon in the BORG toolbar, a web page will be launched asking you to grant access to the new google app that you created (whatever you called it when you created your google project). You probably will have to go past warnings that say that the app is unknown or untrusted. You eventually get to a google screen where you grant both calendar and task permissions to BORG. Once you get past this, then BORG should sync with google.


If this all doesn't work, then just start a discussion on github or email me at mike@mbcsoft.com

BORG and Google do not exactly match up. You can create anything in BORG - like repeating tasks - and something will show up in google. However, the reverse is not always true. You certainly can create/update appointments and non-recurring tasks on your phone/web and have them appear in BORG when you sync. For questions about the sync details start a discussion here or email me.
